certora/specs/ConsistentState.spec

```diff
@@ -155,5 +170,92 @@ function hasSupplyCapIsEnabled(MetaMorphoHarness.Id id) returns bool {
     return supplyCap > 0 => enabled;
 }
 
+// Check that having a positive supply cap implies that the market is enabled.
+// This invariant is useful to conclude that market that are not enabled cannot be interacted with (notably for reallocate).
 invariant supplyCapIsEnabled(MetaMorphoHarness.Id id)
     hasSupplyCapIsEnabled(id);
+
+function hasDistinctIdentifiers(uint256 i, uint256 j) returns bool {
+    return i != j => withdrawQueue(i) != withdrawQueue(j);
+}
+
+// Check that there are no duplicate markets in the withdraw queue.
+invariant distinctIdentifiers(uint256 i, uint256 j)
+    hasDistinctIdentifiers(i, j)
+{
+    preserved updateWithdrawQueue(uint256[] indexes) with (env e) {
+        require hasDistinctIdentifiers(indexes[i], indexes[j]);
```

Comment:
This is the induction hypothesis, we don't need to verify it. I made that more clear by changing it into a `requireInvariant`

Reply:
which induction sorry? what i understand here is that we want to prove that hasDistinctIdentifier is an invariant, but we assume that updateWithdrawQueue (which is touching the identifiers) doesn't break this invariant.

Reply:
An invariant is a property that is always true. It is proven by induction: prove that it's true at deployment (base case), and prove that each function does not change that invariant (induction step). The `preserved` block is allowing us to evaluate some Certora code **before** the corresponding call (here `updateWithdrawQueue`). Since this is an induction step, it's ok to assume that the property is true before doing the call. Here it's useful because the prover would not have guessed by itself that the arguments it needs to pass are `indexes[i]` and `indexes[j]`, by default it tries with the same parameters, that is `i` and `j` directly

Reply:
ok, I didn't remember exactly how to use "preserved blocks". so it's also proved for all functions at the same time right (but not with this precise assumption, just the general assumption)?

Reply:
Yes it proves for all the other functions without issues with the general assumption `hasDistinctIdentifiers(i, j)`. For `updateWithdrawQueue` the prover needed some extra help, and this preserved block only applies to that case